object/datastore_file_manager.go

```diff
@@ -0,0 +1,144 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package object
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"fmt"
+	"path"
+	"strings"
+
+	"github.com/vmware/govmomi/vim25/soap"
+)
+
+// DatastoreFileManager combines FileManager and VirtualDiskManager to manage files on a Datastore
+type DatastoreFileManager struct {
+	Datacenter         *Datacenter
+	Datastore          *Datastore
+	FileManager        *FileManager
+	VirtualDiskManager *VirtualDiskManager
+
+	Force bool
+}
+
+// NewFileManager creates a new instance of DatastoreFileManager
+func (d Datastore) NewFileManager(dc *Datacenter, force bool) *DatastoreFileManager {
+	c := d.Client()
+
+	m := &DatastoreFileManager{
+		Datacenter:         dc,
+		Datastore:          &d,
+		FileManager:        NewFileManager(c),
+		VirtualDiskManager: NewVirtualDiskManager(c),
+		Force:              force,
+	}
+
+	return m
+}
+
+// Delete dispatches to the appropriate Delete method based on file name extension
+func (m *DatastoreFileManager) Delete(ctx context.Context, name string) error {
+	switch path.Ext(name) {
+	case ".vmdk":
+		return m.DeleteVirtualDisk(ctx, name)
+	default:
+		return m.DeleteFile(ctx, name)
+	}
+}
+
+// DeleteFile calls FileManager.DeleteDatastoreFile
+func (m *DatastoreFileManager) DeleteFile(ctx context.Context, name string) error {
+	p := m.Path(name)
+
+	task, err := m.FileManager.DeleteDatastoreFile(ctx, p.String(), m.Datacenter)
+	if err != nil {
+		return err
+	}
+
+	return task.Wait(ctx)
+}
+
+// DeleteVirtualDisk calls VirtualDiskManager.DeleteVirtualDisk
+// Regardless of the Datastore type, DeleteVirtualDisk will fail if 'ddb.deletable=false',
+// so if Force=true this method attempts to set 'ddb.deletable=true' before starting the delete task.
+func (m *DatastoreFileManager) DeleteVirtualDisk(ctx context.Context, name string) error {
+	p := m.Path(name)
+
+	if m.Force {
+		m.markDiskAsDeletable(ctx, p)
+	}
+
+	task, err := m.VirtualDiskManager.DeleteVirtualDisk(ctx, p.String(), m.Datacenter)
+	if err != nil {
+		return err
+	}
+
+	return task.Wait(ctx)
+}
+
+// Path converts path name to a DatastorePath
+func (m *DatastoreFileManager) Path(name string) *DatastorePath {
+	var p DatastorePath
+
+	if !p.FromString(name) {
+		p.Path = name
+		p.Datastore = m.Datastore.Name()
+	}
+
+	return &p
+}
+
+func (m *DatastoreFileManager) markDiskAsDeletable(ctx context.Context, path *DatastorePath) {
+	r, n, err := m.Datastore.Download(ctx, path.Path, &soap.DefaultDownload)
+	if err != nil {
+		return
+	}
+
+	defer r.Close()
+
+	if n > 2048 {
+		return // sanity check before reading; should be only a few hundred bytes
```

Comment:
This is true for the moment but we are discussing using the ddb to store the layer metadata. Regardless, if we've somehow accidentally downloaded the data file instead of the descriptor this will be Gigabytes too late - we could download only the first 512 bytes and check for known keys (for example) - not really thought about edge cases with this suggestion however.